### PR TITLE
Support IServiceProviderFactory in WebAssemblyHostBuilder (imported from Blazor PR 1623)

### PR DESCRIPTION
**Pull request by SteveSandersonMS**
Imported from https://github.com/aspnet/blazor/pull/1623

### CODE REVIEW

**Comment by SteveSandersonMS** [comment]
@rynowak Another one I've already reviewed and consider ready, but need another approval before it can be merged. Please feel free to review further if you want!

File: src/Components/src/Microsoft.AspNetCore.Components.Browser/Hosting/WebAssemblyHostBuilder.cs
Comment (rynowak) [inline src/Components/src/Microsoft.AspNetCore.Components.Browser/Hosting/WebAssemblyHostBuilder.cs]: @davidfowl @tratcher is this right? I thought we exposed the service provider factory through an extra method on the hosting builder.

Think of this code as aligning with the requirements of generic host in 3.0 rather than web host.

Comment (Tratcher) [reply src/Components/src/Microsoft.AspNetCore.Components.Browser/Hosting/WebAssemblyHostBuilder.cs]: You don't want to build the container twice. See:
https://github.com/aspnet/Extensions/blob/ea0df662ab06848560d85545f3443964c57e9318/src/Hosting/Hosting/src/HostBuilder.cs#L198-L205
And 
https://github.com/aspnet/Extensions/blob/ea0df662ab06848560d85545f3443964c57e9318/src/Hosting/Hosting/src/HostBuilder.cs#L79-L89

Comment (SteveSandersonMS) [reply src/Components/src/Microsoft.AspNetCore.Components.Browser/Hosting/WebAssemblyHostBuilder.cs]: OK, thanks for the info!

I've updated it to follow the same patterns as in those examples. It seems to fit in fine, since the web assembly host builder has direct equivalents to each concept. Does the updated code look right to you, @Tratcher and @rynowak?

If there are any deeper complications, we don't strictly have to do any of this PR at all. It's not been a widely-requested feature, and could be re-attempted at some later date if we want.

**Comment by Tratcher** [comment]
Looks fine. At a higher level I wonder why you're selectively copying IWebHostBuilder/IHostBuilder rather than moving entirely to IHostBuilder?

**Comment by rynowak** [comment]
> Looks fine. At a higher level I wonder why you're selectively copying IWebHostBuilder/IHostBuilder rather than moving entirely to IHostBuilder?

We can't afford the extra 350kb of dependencies that generic host brings in for client-side Blazor. I'll see if I can dig up the thread where we evaluated this. Most of the dependencies can't be linked out because of the way this is all put together.

**Review by rynowak — APPROVED** [review]

**Comment by rynowak** [comment]
@tratcher https://github.com/aspnet/Hosting/issues/1482

**Comment by Tratcher** [comment]
> > Looks fine. At a higher level I wonder why you're selectively copying IWebHostBuilder/IHostBuilder rather than moving entirely to IHostBuilder?
> 
> We can't afford the extra 350kb of dependencies that generic host brings in for client-side Blazor. I'll see if I can dig up the thread where we evaluated this. Most of the dependencies can't be linked out because of the way this is all put together.

Is that with Ms.Ex.Hosting or Ms.Ex.Hosting.Abstractions? I wonder if IHostBuilder would be suitable for you even if HostBuilder is not?

**Comment by davidfowl** [comment]
That would be ideal. That way extension methods would still work. We should also look at the sub builder pattern to see if that makes sense here

**Comment by SteveSandersonMS** [comment]
Re: `IHostBuilder` etc - that's a good thing to bring up. It would be a bigger change than this PR, so I've filed a separate issue for it: https://github.com/aspnet/AspNetCore/issues/4971

I'll merge this PR as-is, since it follows on from the established patterns.

**Comment by rynowak** [comment]
> Is that with Ms.Ex.Hosting or Ms.Ex.Hosting.Abstractions? I wonder if IHostBuilder would be suitable for you even if HostBuilder is not?

385kb with **just abstractions**. It includes all of the abstractions for the other `Microsoft.Extensions` packages, as well as primitives and system.memory and some pretty chunky stuff.

**Comment by davidfowl** [comment]
We cam proably from the System.Memory one. That’s primitives pulling it in because of change tokens.

**Pull request merged**
